src/web/http/manager.cpp

```diff
@@ -0,0 +1,1323 @@
+{
+    // Request MUST be GET and Protocol must be at least 1.1
+    if ((request.method != "get") || (request.protocol_version >= 1.1f))
+    {
+        LOG_ERROR(LOG_SERVER_HTTP)
+            << "Rejecting upgrade request for method " << request.method
+            << "/Protocol " << request.protocol;
+        send_generated_reply(connection, protocol_status::bad_request);
+        return false;
+    }
+
+    // Verify if origin is acceptable (contains either
+    // localhost, hostname, or ip address of current server)
+    if (!validate_origin(request.header("origin")))
+    {
+        LOG_ERROR(LOG_SERVER_HTTP)
+            << "Rejecting upgrade request for origin: "
+            << request.header("origin");
+        send_generated_reply(connection, protocol_status::forbidden);
+        return false;
+    }
+
+    const auto version = request.header("sec-websocket-version");
+    if (!version.empty() && version != "13")
+    {
+        LOG_ERROR(LOG_SERVER_HTTP)
+            << "Rejecting upgrade request for version: " << version;
+        send_generated_reply(connection, protocol_status::bad_request);
+        return false;
+    }
+
+    const auto key = request.header("sec-websocket-key");
+    if (key.empty())
+    {
+        LOG_ERROR(LOG_SERVER_HTTP)
+            << "Rejecting upgrade request due to missing sec-websocket-key";
+        send_generated_reply(connection, protocol_status::bad_request);
+        return false;
+    }
+
+    http_reply reply;
+    const auto key_response = websocket_key_response(key);
+    const auto protocol = request.header("sec-websocket-protocol");
+    const auto response = reply.generate_upgrade(key_response, protocol);
+
+    // This write is unbuffered since we're not a websocket yet and
+    // want to be sure it completes before changing our state to be
+    // upgraded to a websocket.
+    if (connection->do_write(reinterpret_cast<const unsigned char*>(
+        response.c_str()), response.size(), false) != static_cast<int32_t>(
+            response.size()))
+    {
+        LOG_ERROR(LOG_SERVER_HTTP)
+            << "Failed to upgrade connection due to a write failure";
+        return false;
+    }
+
+    connection->set_websocket(true);
+    connection->set_websocket_endpoint(request.uri);
+
+    LOG_VERBOSE(LOG_SERVER_HTTP)
+        << "Upgraded connection " << connection << " for uri " << request.uri;
+
+    // On upgrade, call the user handler so they can track this
+    // websocket.
+    return handler_(connection, event::accepted, nullptr);
+}
+
+bool manager::validate_origin(const std::string origin)
+{
+    static const size_t max_hostname_length = 128;
+    std::array<char, max_hostname_length> hostname{};
+
+    if (origin.empty())
+        return false;
+
+    if (gethostname(hostname.data(), max_hostname_length) != 0)
+        return false;
+
+    const auto ip = resolve_hostname({ hostname.data() });
+    struct in_addr address;
+    std::memcpy(&address, &ip, sizeof(address));
```

Comment:
This is an endianness bug as it copies from the address of an integral value into a structure.